### PR TITLE
Refactor node_repositories

### LINKED ISSUE #181: Using non-built-in versions from scripts calling `npm`/`yarn`
Assuming the following "npm scripts" in `package.json`:

```json
"scripts": {
  "npm-version": "npm -v",
  "yarn-version": "yarn -v",
  ...
```

1. `bazel run @nodejs//:npm run yarn-version` will not use the built-in `yarn`. (This might or might not be by design and the usecase is rather uncommon - so not too worried about this :grin:)

2. `bazel run @yarn//:yarn run yarn-version` will not use the built-in `yarn`. This is quite unexpected.

3. `bazel run @yarn//:yarn run npm-version` will not use the built-in `npm`. This is also quite unexpected.

Related to #140.

### PR DESCRIPTION
**Pull request by gregmagolan**
Simplifies code & fixes #181.

Ability to specify version of node & yarn as well as link to local packages is now baked into node_repositories() function and no repository rules are exposed from node_repositories.bzl.

Custom repositories & URLs for download can be configured via node_repositories() parameters.

Yarn is now downloaded to @nodejs workspace. `@nodejs//:yarn` is now the default and recommended target. `@yarn//:yarn` is still available for backward compat.

nodejs repository folder `repository_bin` is put on the PATH when running node. This path includes entry points for node, npm & yarn. These are `repository_bin/node`, `repository_bin/npm`, `repository_bin/yarn` for osx/linux && `repository_bin/node.cmd`, `repository_bin/npm.cmd`, `repository_bin/yarn.cmd` for windows. These entry points handle adding repository_bin to the PATH.

### CODE REVIEW

File: internal/node/node_repositories.bzl [hunk 1]
```diff
@@ -95,297 +163,174 @@ def _prepare_node(repository_ctx):
...
 #Generated by node_repositories.bzl
+#Executes the given NPM command over each of the package.json folders provided in node_respositories.
 """ + "".join(["""
-NODE="{}"
-NODE_PATH="{}"
-SCRIPT="{}"
-ROOT="{}"
-export PATH="$NODE_PATH":$PATH
-"$NODE" "$SCRIPT" --cwd "$ROOT" "$@"
+"{}" "{}" --cwd "{}" "$@"
```
Comment (alexeagle) [inline internal/node/node_repositories.bzl]: again probably should be named placeholders

File: internal/node/node_repositories.bzl [hunk 2]
```diff
@@ -442,45 +402,33 @@ def node_repositories(package_json, node_version=DEFAULT_NODE_VERSION, yarn_vers
...
+    name = "build_bazel_rules_nodejs_npm_install_deps",
+    package_json = "@build_bazel_rules_nodejs//internal/npm_install:package.json",
+    yarn_lock = "@build_bazel_rules_nodejs//internal/npm_install:yarn.lock",
   )
 
   yarn_install(
-      name = "build_bazel_rules_nodejs_rollup_deps",
-      package_json = "@build_bazel_rules_nodejs//internal/rollup:package.json",
-      yarn_lock = "@build_bazel_rules_nodejs//internal/rollup:yarn.lock",
+    name = "build_bazel_rules_nodejs_rollup_deps",
```
Comment (alexeagle) [inline internal/node/node_repositories.bzl]: nit: 4 space indent is correct for continuation

File: README.md [hunk 3]
```diff
@@ -53,12 +53,18 @@ load("@build_bazel_rules_nodejs//:defs.bzl", "node_repositories")
...
 ```
 
-#### Currently supported versions
+#### Currently supported versions of NodeJS
 
 * 9.11.1
 * 8.11.1
 * 8.9.1
 
+#### Currently supported versions of NodeJS
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply README.md]: Thanks

File: internal/node/node_repositories.bzl (hunk 2 above)
Comment (gregmagolan) [reply internal/node/node_repositories.bzl]: 👍

File: internal/node/node_repositories.bzl [hunk 4]
```diff
@@ -95,297 +163,174 @@ def _prepare_node(repository_ctx):
...
   host = _get_host(repository_ctx)
-
-  if host.find("windows") != -1:
-    # The windows distribution of nodejs has the binaries in different paths
-    node = "node.exe"
-    npm = "node_modules/npm/bin/npm-cli.js"
-  else:
-    node = "bin/node"
-    npm = "bin/npm"
+  node_exec = "node/bin/node" if host.find("windows") == -1 else "node/node.exe"
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/node/node_repositories.bzl]: Good idea

File: internal/node/node_repositories.bzl (hunk 1 above)
Comment (gregmagolan) [reply internal/node/node_repositories.bzl]: Good idea

**Comment by gregmagolan** [comment]
@chrisleck I did some refactoring in node_respositories.bzl here to fix #181 and remove some of the duplication. The functionality you put in is still all there but it is now all configured via parameters to the node_repositories() function.

File: README.md [hunk 5]
```diff
@@ -71,33 +77,28 @@ git_repository(
...
+  yarn_version = "1.5.1",
+  node_repositories = {
+    "8.10.0-darwin_amd64": ("node-v8.10.0-darwin-x64.tar.gz", "node-v8.10.0-darwin-x64", "7d77bd35bc781f02ba7383779da30bd529f21849b86f14d87e097497671b0271"),
+    "8.10.0-linux_amd64": ("node-v8.10.0-linux-x64.tar.xz", "node-v8.10.0-linux-x64", "92220638d661a43bd0fee2bf478cb283ead6524f231aabccf14c549ebc2bc338"),
+    "8.10.0-windows_amd64": ("node-v8.10.0-win-x64.zip", "node-v8.10.0-win-x64", "936ada36cb6f09a5565571e15eb8006e45c5a513529c19e21d070acf0e50321b"),
+  },
+  yarn_repositories = {
+    "1.5.1": ("yarn-v1.5.1.tar.gz", "yarn-v1.5.1", "cd31657232cf48d57fdbff55f38bfa058d2fb4950450bd34af72dac796af4de1"),
+  },
+  node_urls = ["https://nodejs.org/dist/v{version}/{filename}"],
```
Comment (alexeagle) [inline README.md]: at least for the "supported" versions we ought to offer a second mirror for downloads, from the mirror.bazel.build

Comment (gregmagolan) [reply README.md]: Good idea

**Comment by gregmagolan** [comment]
Local node & yarn repositories working on Windows now

**Comment by chrisleck** [comment]
Thanks for the heads up.  Seems to work locally for me.

File: README.md (hunk 5 above)
Comment (alexeagle) [reply README.md]: Doing the mirroring now

File: internal/node/node_repositories.bzl [hunk 6]
```diff
@@ -28,21 +28,18 @@ DEFAULT_YARN_VERSION = "1.3.2"
 
 # Dictionary mapping NodeJS versions to sets of hosts and their correspoding (filename, strip_prefix, sha256) tuples.
 NODE_REPOSITORIES = {
```
Comment (alexeagle) [inline internal/node/node_repositories.bzl]: Mirrored by running 
```
$ for v in 9.11.1 8.11.1 8.9.1; do for plat in darwin-x64.tar.gz linux-x64.tar.xz win-x64.zip; do bzmirror https://nodejs.org/dist/v${v}/node-v${v}-${plat}; done; done
```

File: internal/node/node_repositories.bzl [hunk 7]
```diff
@@ -52,6 +49,10 @@ YARN_REPOSITORIES = {
   "1.3.2": ("yarn-v1.3.2.tar.gz", "yarn-v1.3.2", "6cfe82e530ef0837212f13e45c1565ba53f5199eec2527b85ecbcd88bf26821d"),
 }
 
+# Urls patterns for downloading node & yarn distributions
+NODE_URLS = ["https://nodejs.org/dist/v{version}/{filename}"]
+YARN_URLS = ["https://github.com/yarnpkg/yarn/releases/download/v{version}/{filename}"]
```
Comment (alexeagle) [inline internal/node/node_repositories.bzl]: mirrored weth 
```
for v in 1.6.0 1.5.1 1.3.2; do bzmirror https://github.com/yarnpkg/yarn/releases/download/v${v}/yarn-v${v}.tar.gz; done
```

**Comment by alexeagle** [comment]
Could you update this one to use the Bazel mirror URL first for node.js and yarn downloads on any of the supported versions? they should all be mirrored now.

**Pull request merged**